File: src/main/java/com/toolforge/service/UserService.java
package com.toolforge.service;

import com.toolforge.model.User;
import com.toolforge.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Optional;

/**
 * Handles user business logic — mirrors the logic in POST /api/signup and Passport LocalStrategy.
 */
@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Register a new user. Throws IllegalArgumentException on validation failures.
     */
    public User register(String name, String username, String email,
                         String password, String confirmPassword) {
        if (!password.equals(confirmPassword)) {
            throw new IllegalArgumentException("Passwords do not match.");
        }
        if (userRepository.existsByEmail(email.toLowerCase())) {
            throw new IllegalArgumentException("Email already registered.");
        }
        if (userRepository.existsByUsername(username.toLowerCase())) {
            throw new IllegalArgumentException("Username already taken.");
        }

        User user = new User(name, username.toLowerCase(), email.toLowerCase(),
                passwordEncoder.encode(password));
        return userRepository.save(user);
    }

    /**
     * Authenticate a user by username + password.
     * Returns empty Optional if credentials are wrong.
     */
    public Optional<User> authenticate(String username, String rawPassword) {
        Optional<User> userOpt = userRepository.findByUsername(username.toLowerCase());
        if (userOpt.isEmpty()) return Optional.empty();

        User user = userOpt.get();
        if (user.getPassword() == null || user.getPassword().isEmpty()) {
            // Google-only account
            return Optional.empty();
        }
        if (!passwordEncoder.matches(rawPassword, user.getPassword())) {
            return Optional.empty();
        }
        return Optional.of(user);
    }

    /**
     * Find or create a user from Google OAuth profile.
     * Mirrors the GoogleStrategy verify callback.
     */
    public User findOrCreateGoogleUser(String googleId, String email,
                                        String displayName, String avatar) {
        // Already linked by Google ID
        Optional<User> byGoogleId = userRepository.findByGoogleId(googleId);
        if (byGoogleId.isPresent()) return byGoogleId.get();

        // Existing email account — link it
        Optional<User> byEmail = userRepository.findByEmail(email.toLowerCase());
        if (byEmail.isPresent()) {
            User user = byEmail.get();
            user.setGoogleId(googleId);
            user.setAvatar(avatar != null ? avatar : "");
            return userRepository.save(user);
        }

        // Brand new user
        String baseUsername = displayName.toLowerCase()
                .replaceAll("\\s+", "_")
                .replaceAll("[^a-z0-9_]", "");
        String finalUsername = baseUsername;
        int count = 1;
        while (userRepository.existsByUsername(finalUsername)) {
            finalUsername = baseUsername + count++;
        }

        User user = new User();
        user.setGoogleId(googleId);
        user.setName(displayName);
        user.setUsername(finalUsername);
        user.setEmail(email.toLowerCase());
        user.setAvatar(avatar != null ? avatar : "");
        return userRepository.save(user);
    }

    public Optional<User> findById(String id) {
        return userRepository.findById(id);
    }
}

File: app/services/user_service.py
"""
User service for handling registration and authentication.
"""

from typing import Optional
import bcrypt
from app.models.user import User


async def authenticate(username: str, password: str) -> Optional[User]:
    """
    Verify username/password.
    Mirrors UserService.authenticate() — uses BCrypt comparison.
    """
    user = await User.find_one(User.username == username.lower())
    if not user or not user.password:
        return None
    if bcrypt.checkpw(password.encode(), user.password.encode()):
        return user
    return None


async def register(
    name: str,
    username: str,
    email: str,
    password: str,
    confirm_password: str,
) -> User:
    """
    Create new user account.
    Mirrors UserService.register() with validation.
    """
    if password != confirm_password:
        raise ValueError("Passwords do not match")

    username = username.lower()
    email = email.lower()

    existing = await User.find_one(
        (User.username == username) | (User.email == email)
    )
    if existing:
        if existing.username == username:
            raise ValueError("Username already taken")
        raise ValueError("Email already registered")

    hashed = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
    user = User(name=name, username=username, email=email, password=hashed)
    await user.insert()
    return user


def to_user_response(user: User) -> dict:
    return {
        "id": str(user.id),
        "name": user.name,
        "username": user.username,
        "email": user.email,
        "avatar": user.avatar,
        "google_id": user.google_id,
    }


async def get_by_email(email: str) -> Optional[User]:
    return await User.find_one(User.email == email.lower())


async def create_oauth_user(email: str, name: str, google_id: str, avatar: str) -> User:
    username = email.split("@")[0].lower()
    existing = await User.find_one(User.username == username)
    if existing:
        username = f"{username}_{google_id[:5]}"

    user = User(
        name=name,
        username=username,
        email=email.lower(),
        google_id=google_id,
        avatar=avatar,
    )
    await user.insert()
    return user
